File: 2016/day12.cpp
// https://adventofcode.com/2016/day/12

#include "assembunny.h"

#ifndef AOC_IMPORT_STD
#include <print>
#endif

using assembunny::Registers;
using assembunny::XOp;

using Input = Vec<XOp>;

auto parse(String const& filename) -> Input {
  return assembunny::transform(assembunny::parse(filename));
}

template <i64 RegCInit>
fn solve_case(Input const& ops) -> i64 {
  auto registers = Registers{0, 0, RegCInit, 0};
  auto counter = i64{};
  while ((counter >= 0) && (static_cast<usize>(counter) < ops.size())) {
    let op = ops[static_cast<usize>(counter)];
    (void)assembunny::exec(op, registers, counter);
  }

  // Register a
  return registers[0];
}

int main() {
  std::println("Part 1");
  let example = parse("day12.example");
  AOC_EXPECT_RESULT(42, solve_case<0>(example));
  let input = parse("day12.input");
  AOC_EXPECT_RESULT(318083, solve_case<0>(input));

  std::println("Part 2");
  AOC_EXPECT_RESULT(42, solve_case<1>(example));
  AOC_EXPECT_RESULT(9227737, solve_case<1>(input));

  AOC_RETURN_CHECK_RESULT();
}

File: rust/2016/src/day12.rs
mod assembunny;
use assembunny::{Registers, XOp};

type Input = Vec<XOp>;

fn parse(filename: &str) -> Input {
    return assembunny::transform(&assembunny::parse(filename));
}

fn solve_case<const REG_C_INIT: i64>(ops: &Input) -> i64 {
    let mut registers: Registers = [0, 0, REG_C_INIT, 0];
    let mut counter = 0_i64;
    while ((counter >= 0) && ((counter as usize) < ops.len())) {
        assembunny::exec(&ops[counter as usize], &mut registers, &mut counter);
    }
    // Register a
    return registers[0];
}

fn main() {
    println!("Part 1");
    let example = parse("day12.example");
    aoc::expect_result!(42, solve_case::<0>(&example));
    let input = parse("day12.input");
    aoc::expect_result!(318083, solve_case::<0>(&input));

    println!("Part 2");
    aoc::expect_result!(42, solve_case::<1>(&example));
    aoc::expect_result!(9227737, solve_case::<1>(&input));
}
